Add LNS tutorial - Closes #1151 (#1158)
LiskHQ/lisk-docs: modules/ROOT/pages/tutorials/lns/1-blockchain.adoc

// Project URLs
:url_guide_dashboard: guides/app-development/dashboard.adoc
:url_guide_setup_dependencies: guides/app-development/setup.adoc#dependencies
:url_guide_setup_init: guides/app-development/setup.adoc#initializing-a-new-blockchain-application
:url_guide_genesisblock: guides/app-development/genesis-block.adoc
:url_reference_dashboard: references/lisk-framework/dashboard-plugin.adoc
:url_reference_faucet: references/lisk-framework/faucet-plugin.adoc
:sectnums:

== Dependencies

Please refer to the section xref:{url_guide_setup_dependencies}[Lisk SDK Dependencies] of the setup guide to set up all the required dependencies to follow this tutorial, if you haven't done so alrerady.

== Bootstrapping the default application

Please follow the section xref:{url_guide_setup_init}[Initializing a new blockchain application] of the setup guide to bootstrap the LNS blockchain application.

[NOTE]
====
Please use `lisk-name-service/lns` as root folder for the LNS blockchain application, and `lns` as application name.

[source,bash]
----
mkdir lisk-name-service
cd lisk-name-service
lisk init
----
====

> To customize the default blockchain application to suit our desired use case, we will now generate the LNS module skeleton.

Author: ```suggestion
To customize the default blockchain application to suit our desired use case, generate the LNS module skeleton.
```

    type: 'object',
    required: [],
    properties: {},
  };

  public validate({ asset }: ValidateAssetContext<{}>): void {
    // Validate your asset
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  public async apply({ asset, transaction, stateStore }: ApplyAssetContext<{}>): Promise<void> {
    throw new Error('Asset "register" apply hook is not implemented.');
  }
}
----

As you can see, the asset name and ID are already pre-filled with the values we provided when creating the asset skeleton.

As next step, we want to define the asset schema, which defines which kind of data is expected by the application to successfully register a new domain for a user account.

==== Schema

Create a new folder `data/` inside the `lns` module folder.

.src/app/modules/lns/
[source,bash]
----
mkdir data
----

This folder is created to maintain a better overview, and will store all account and asset schemas which are relevant to the LNS module.

.src/app/modules/lns/
[source,bash]
----
mkdir data/assets
----

Inside the `data/assets` folder, create a new file `register.ts`, which will contain the schemas related to the `register` asset.

> The first thing we define in the file, is an interface  for the expected asset data of a `register` transaction.

Author: ```suggestion
Firstly define an interface for the expected asset data of a `register` transaction.
```

    ttl: {
      dataType: 'uint32',
      fieldNumber: 2,
    },
    registerFor: {
      dataType: 'uint32',
      fieldNumber: 3,
    },
  },
}
----

Add the interface and asset schema to the file and save it.

Now,  just include the schema in the asset file:

.src/app/modules/lns/assets/register.ts
[source,typescript]
----
import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';
import { RegisterAssetProps, registerAssetPropsSchema } from '../data';

export class RegisterAsset extends BaseAsset<RegisterAssetProps> {
  public name = 'register';
  public id = 1;

  // Define schema for asset
  public schema = registerAssetPropsSchema;

  // ...
}
----

==== Validation

Create a new file `constants.ts` inside the `lns` module folder.

This file is created to maintain a better overview, and will store all constants which are relevant to the LNS module and its' assets.

Add the following two constants.

.src/app/modules/lns/constants.ts
[source,typescript]
----
export const MIN_TTL_VALUE = 60 * 60; // 1 hour
export const VALID_TLDS = ['lsk'];
----

Now import the constants into the `register` asset, and use them to check the validity of transaction assets:

> We want to validate the following:

Author: ```suggestion
Validate the following:
```

. The TTL value needs to be above the minimum defined TTL value (60 * 60).
. The `registerFor` value needs to be between 1 and 5.
. Only second level domain names can be registered.
. Only domains with valid TLDs can be registered.

> The corresponding checks look like this:

Author: ```suggestion
The corresponding code looks like this:
```

import * as namehash from 'eth-ens-namehash';

// ...

export const getLNSObject = async (
	stateStore: StateStore,
	node: Buffer,
): Promise<LNSNode | undefined> => {
	const result = await stateStore.chain.get(getKeyForNode(node));

	if (!result) {
		return;
	}

	// eslint-disable-next-line consistent-return
	return codec.decode<LNSNode>(lnsNodeSchema, result);
};
----

==== Schemas
Implement the following interfaces and schemas, which are be used inside the utility functions:

* <<lns-node-record-schema>>: Interface and schema for an LNS node record.
* <<lns-node-schema>>: Interface and schema for an LNS node.

===== LNS node record schema

Create a new file `data/lns_node_records.ts` in the LNS module folder.

The interface for an LNS node record looks like this:

.src/app/modules/lns/data/lns_node_records.ts
[source,typescript]
----
export interface LNSNodeRecord {
	type: number; // <1>
	label: string; // <2>
	value: string; // <3>
}

export type LNSNodeRecordJSON = LNSNodeRecord;
----

<1> `type`: Type of the records as number.
`1` stands for a CNAME record, `2` stands for a TXT record.
<2> `label`: Label for the record.
<3> `value`: Value for the record.

> Based on this interface, we can create the corresponding schema, which looks like this:

Author: ```suggestion
Based on this interface, create the corresponding schema, as shown below:
```

* Add the hash value of the domain as the reverse lookup domain to the senders account under the key `lns.reverseLookup`.

.src/app/modules/lns/assets/reverse_lookup.ts
[source,typescript]
----
public async apply({
    asset,
    stateStore,
    transaction,
}: ApplyAssetContext<ReverseLookupAssetProps>): Promise<void> {
    const node = getNodeForName(asset.name);
    const sender = await stateStore.account.get<LNSAccountProps>(transaction.senderAddress);

    const exists = sender.lns.ownNodes.find(n => n.equals(node));

    if (!exists) {
        throw new Error('You can only assign lookup node which you own.');
    }

    sender.lns.reverseLookup = node;
    await stateStore.account.set(sender.address, sender);
}
----

=== Update records

Finally, create the asset for updating the records of a domain.
This is the last of the three assets in the LNS module.

[NOTE]

The {url_wikipedia_cname}[CNAME^] and {url_wikipedia_txt}[TXT^] records have no distinct difference in the LNS blockchain application.
Different handling of the records can be implemented at a later stage.

> I.e. if you build any DNS provider service on top of the LNS app, then you can use CNAME and TXT records differently there.

Author: ```suggestion
For example if you build any DNS provider service on top of the LNS app, then you can use CNAME and TXT records differently there.
```

. <<assets>>
. <<actions>>
. <<reducers>>

The events and lifecycle hooks are not required in this use case, so you can leave the skeletons as they are.

=== Assets

Now, let's add the assets we created before in step <<creating-the-lns-module-assets>>.
Import the different assets to the LNS module as shown in the snippet below.

Then, create a new instance of each asset and add them as array to the `transactionAssets` property of the LNS module.

.lns/lns_module.ts
[source,typescript]
----
import { RegisterAsset } from './assets/register';
import { UpdateRecordsAsset } from './assets/update_records';
import { ReverseLookupAsset } from './assets/reverse_lookup';

export class LnsModule extends BaseModule {
    // ...
    public transactionAssets = [
		new RegisterAsset(),
		new ReverseLookupAsset(),
		new UpdateRecordsAsset(),
	];
    // ...
}
----

That's all that is needed to add new assets to the LNS module.

=== Actions

The LNS module should have the following actions:

* `lookupAddress`: returns an LNS object based on a provided account address.
* `resolveName`: returns an LNS object based on a domain name.
* `resolveNode`: returns an LNS object based on a node hash value.

> All three actions are returning an <<lns-node-schema, LNS object>>, based on different input parameters like the address it is registered to, the domain name that is registered in the LNS object, or a hash of the LNS object, which is typically stored in user account under the key `lns.ownNodes`.

Author: ```suggestion
All three actions are returning an <<lns-node-schema, LNS object>>, based on different input parameters, like: the address it is registered to, the domain name that is registered in the LNS object, or a hash of the LNS object, which is typically stored in user account under the key `lns.ownNodes`.
```

		throw new Error(`Account "${address.toString('hex')}" is not associated with any LNS object.`);
	}

	const result = await chainGetter(getKeyForNode(account.lns.reverseLookup));

	if (!result) {
		throw new Error(`Problem looking up node "${account.lns.reverseLookup.toString('hex')}"`);
	}

	const lnsNode = codec.decode<LNSNode>(lnsNodeSchema, result);

	if (isExpired(lnsNode)) {
		throw new Error(`Account "${address.toString('hex')}" is associated to an expired LNS object.`);
	}

	return lnsNode;
};

// ...
----

With this last step, all required parts of the LNS module are implemented.

=== Updating the genesis block

NOTE: In case the LNS tutorial was cloned from the `lisk-sdk-examples` repository, this step can be skipped, as the example application already contains the correct genesis block.

If the blockchain application was bootstrapped newly with `lisk init`, it still contains a default genesis block, which doesnt include the account schema of the new LNS module.
Therefore it is needed to create a new genesis block with genesis accounts that include the new account properties of the LNS module.

//TODO: Add more information about how to update the genesis block

> TIP: A lot of detailed information about this process can be found in the guide xref:{url_guide_genesisblock}[].

Author: ```suggestion
TIP: Detailed information about this process can be found in the guide xref:{url_guide_genesisblock}[].
```

./lisk-name-service/lns/
[source,bash]
----
./bin/run genesis-block:create --output  /home/USERNAME/.lisk/lns/config/default
----

Copy the following content to `~/.lisk/lns/config/default/config.json`:

* copy the content of `forging_info.json` to under forging.delegates
* copy the content of `password.json` to under forging.defaultPassword

Save the file `accounts.json` somewhere safe, it contains the credentials for all genesis accounts.
In the LNS example application, the file is stored in the {url_sdkexamples_lns_secret}[lisk-name-service/lns/.secret/^] folder.

After the genesis block and config are updated, the application should start again successfully. To start, run the following command:

.lisk-name-service/lns/
[source,bash]
----
./bin/run start
----

> Before we go on with developing the frontend part of the LNS application, let's quickly check if the blockchain application is working as expected by enabling the Dashboard plugin in the next chapter.

Author: ```suggestion
Before proceeding with developing the frontend part of the LNS application, let's quickly check if the blockchain application is working as expected by enabling the Dashboard plugin in the next chapter.
```

** ... perform a reverse lookup for an account address.

Before the Dashboard and Faucet plugins can be used for this purpose, it is first needed to install them and to register them with the LNS application, which is done in the next step.

[NOTE]
====
More information about the Dashboard plugin can be found in the development guide xref:{url_guide_dashboard}[] or on the reference page of the xref:{url_reference_dashboard}[].

More information about the Faucet plugin can be found on the reference page of the xref:{url_reference_faucet}[].
====

=== Registering the Dashboard plugin

Install the dashboard plugin:

.lisk-name-service/
[source,bash]
----
npm i @liskhq/lisk-framework-dashboard-plugin
----

Now open `plugins.ts`, import the Dashboard plugin, and register it with the application as shown below:

.lisk-name-service/lns/src/app/plugins.ts
[source,typescript]
----
import { Application } from 'lisk-sdk';
import { DashboardPlugin } from "@liskhq/lisk-framework-dashboard-plugin";

export const registerPlugins = (app: Application): void => {

    app.registerPlugin(DashboardPlugin);
};
----

Save and close `plugins.ts`.

=== Registering the Faucet plugin

[TIP]

> You can skip this step, if you don't create a new account in step <<create-new-account>>, but rather use on of the existing genesis accounts.

Author: ```suggestion
This step can be skipped if you do not create a new account in the step <<create-new-account>>, and rather use one of the existing genesis accounts instead.
```

The {url_recaptcha_keys}[free site key and secret key for reCAPTCHA^] are used below for testing purposes.

.~/.lisk/lns/config/default/config.json
[source,json]
----
"plugins": {
    "faucet": {
        "encryptedPassphrase": "iterations=1000000&cipherText=643bfbf1b6f1dc0ce740dd9fc9f27a682e476dc5de4e6c023deded4d3efe2822346226541106b42638db5ba46e0ae0a338cb78fb40bce67fdec7abbca68e20624fa6b0d7&iv=8a9c461744b9e70a8ba65edd&salt=3fe00b03d10b7002841857c1f028196e&tag=c57a798ef65f5a7be617d8737828fd58&version=1",
        "captchaSecretkey": "6LeIxAcTAAAAAGG-vFI1TnRWxMZNFuojJ4WifJWe",
        "captchaSitekey": "6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI"
    }
}
----

The last step to use the Faucet plugin is to enable it via the action `faucet:authorize`.
This can be achieved with the Dashboard plugin.

Start the blockchain application again:

[source,bash]
----
./bin/run start
----

Wait until the application start is completed.

Go to {url_dashboard} to access the dashboard.

Now go to the `Call actions` section on the Dashboard, and select the action `faucet:authorize`.

image:tutorials/lns/faucet-authorize-action.png[faucet:authorize,200,100]

> The actions expects as input a boolean, if the plugin should be enabled, and a password the decrypt the encrypted passphrase that was saved in `config.json` above.

Author: ```suggestion
The action expects a boolean  as input, defining if the plugin should be enabled, and also a password to decrypt the encrypted passphrase that was saved in `config.json` above.
```
